util/awsservice/cloudwatchlogs.go

```diff
@@ -161,15 +168,97 @@ func GetLogStreams(logGroupName string) []types.LogStream {
 	return []types.LogStream{}
 }
 
-func MatchEMFLogWithSchema(logEntry string, s *jsonschema.Schema, logValidator func(string) bool) bool {
-	keyErrors, e := s.ValidateBytes(context.Background(), []byte(logEntry))
-	if e != nil {
-		log.Println("failed to execute schema validator:", e)
-		return false
-	} else if len(keyErrors) > 0 {
-		log.Printf("failed schema validation: %v\n", keyErrors)
-		return false
+type LogEventValidator func(event types.OutputLogEvent) error
+
+type LogEventsValidator func(events []types.OutputLogEvent) error
+
+type SchemaRetriever func(message string) (string, error)
+
+func WithSchema(schema string) SchemaRetriever {
+	return func(_ string) (string, error) {
+		return schema, nil
+	}
+}
+
+func AssertLogSchema(schemaRetriever SchemaRetriever) LogEventValidator {
+	return func(event types.OutputLogEvent) error {
+		message := *event.Message
+		if schemaRetriever == nil {
+			return errors.New("nil schema retriever")
+		}
+		schema, err := schemaRetriever(*event.Message)
+		if err != nil {
+			return fmt.Errorf("unable to retrieve schema: %w", err)
+		}
+		keyErrors, err := jsonschema.Must(schema).ValidateBytes(context.Background(), []byte(message))
+		if err != nil {
+			return fmt.Errorf("failed to execute schema validator: %w", err)
+		} else if len(keyErrors) > 0 {
+			return fmt.Errorf("failed schema validation: %v | schema: %s | log: %s", keyErrors, schema, message)
+		}
+		return nil
+	}
+}
+
+func AssertLogSubstring(substr string) LogEventValidator {
```

Comment:
nit: The name did not make it clear if this was asserting the log contains the substring or does not.
We could call it `AssertLogContainsSubstring`